### PR TITLE
Reboot test with USB changes

### PR DESCRIPTION
**Pull request by dautovri**
Added reboot detector 
Test is working on Xen 

https://github.com/lf-edge/eden/pull/546#pullrequestreview-613255388
@eriknordmark @giggsoff  @sadov 

Signed-off-by: Ruslan Dautov <dautov2@gmail.com>

### CODE REVIEW

File: tests/hardware_reboot/testdata/hardware_usb_reboot.txt [hunk 1]
```diff
@@ -13,20 +11,23 @@ exec -t 20m bash ssh.sh 2223
...
+exec -t 20m bash get-lshw.sh 2223
+cp stdout before_reboot
 
 # reboot from guest
 exec -t 20m bash reboot.sh 2223
-exec sleep 10
+exec sleep 60
+
+# reboot detector
+test eden.reboot.test -test.v -timewait 20m -reboot=1
```
Comment (giggsoff) [inline tests/hardware_reboot/testdata/hardware_usb_reboot.txt:24]: It is reboot detector for EVE. And in such notation it will wait for reboot for 20 minutes. Is it expected behavior here?

Comment (zed-sadov) [reply tests/hardware_reboot/testdata/hardware_usb_reboot.txt:24]: Does it seem reasonable to check the USB transfer status in both cases - after stopping/starting the application and also after a full EVE restart?

Comment (eriknordmark) [reply tests/hardware_reboot/testdata/hardware_usb_reboot.txt:24]: The issue we've had in the past was that the handling of a reboot from inside the guest failed to re-create the device passthrough. That is a completely different code path at least in Xen than the normal stop/start of the guest.

Rebooting the device and starting the application/guest a second later isn't any different than booting the device and starting the application/guest ten minutes later. So testing that isn't critical.

Comment (sadov) [reply tests/hardware_reboot/testdata/hardware_usb_reboot.txt:24]: OK, @eriknordmark  just a final clarification -- we need a test that passes on EVE-xen with checking the USB-passthrough status after rebooting the application. It is right?

**Comment by rvs** [comment]
Hey @sadov @giggsoff what's the latest on this? When do you think we can get this merged?

**Comment by zed-sadov** [comment]
@eriknordmark is it correct?
https://github.com/lf-edge/eden/pull/560#discussion_r596902346

**Comment by zed-sadov** [comment]
An error occurred while compiling:
```
$ git clone https://github.com/dautovri/eden.git
$ go mod download github.com/magefile/mage
$ go get github.com/lf-edge/eden/cmd
$ go get github.com/lf-edge/eden/pkg/linuxkit
$ go get golang.org/x/term@v0.0.0-20201210144234-2321bbc49cbf
$ make build-tests
...
make[2]: Entering directory '/data/work/sadov/EVE/github/dautovri/eden/tests/escript'
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go test -c -ldflags "-s -w" -o eden.escript.test-linux-amd64 *.go
# github.com/lf-edge/eden/tests/escript/go-internal/testscript
go-internal/testscript/exe.go:144:37: cannot use nopTestDeps{} (type nopTestDeps) as type testing.testDeps in argument to testing.MainStart:
        nopTestDeps does not implement testing.testDeps (missing SetPanicOnExit0 method)
make[2]: *** [Makefile:52: eden.escript.test-linux-amd64] Error 2
make[2]: Leaving directory '/data/work/sadov/EVE/github/dautovri/eden/tests/escript'
make[1]: *** [Makefile:34: escript/Makefile_build] Error 2
make[1]: Leaving directory '/data/work/sadov/EVE/github/dautovri/eden/tests'
make: *** [Makefile:95: testbin] Error 2
$ go version
go version go1.16.2 linux/amd64
```

**Comment by dautovri** [comment]
> An error occurred while compiling:
> 
> ```
> $ git clone https://github.com/dautovri/eden.git
> $ go mod download github.com/magefile/mage
> $ go get github.com/lf-edge/eden/cmd
> $ go get github.com/lf-edge/eden/pkg/linuxkit
> $ go get golang.org/x/term@v0.0.0-20201210144234-2321bbc49cbf
> $ make build-tests
> ...
> make[2]: Entering directory '/data/work/sadov/EVE/github/dautovri/eden/tests/escript'
> CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go test -c -ldflags "-s -w" -o eden.escript.test-linux-amd64 *.go
> # github.com/lf-edge/eden/tests/escript/go-internal/testscript
> go-internal/testscript/exe.go:144:37: cannot use nopTestDeps{} (type nopTestDeps) as type testing.testDeps in argument to testing.MainStart:
>         nopTestDeps does not implement testing.testDeps (missing SetPanicOnExit0 method)
> make[2]: *** [Makefile:52: eden.escript.test-linux-amd64] Error 2
> make[2]: Leaving directory '/data/work/sadov/EVE/github/dautovri/eden/tests/escript'
> make[1]: *** [Makefile:34: escript/Makefile_build] Error 2
> make[1]: Leaving directory '/data/work/sadov/EVE/github/dautovri/eden/tests'
> make: *** [Makefile:95: testbin] Error 2
> $ go version
> go version go1.16.2 linux/amd64
> ```

Fixed

**Comment by sadov** [comment]
How can this work after `eden eve stop`?
https://github.com/dautovri/eden/blob/bcf2d5a315687c6f080ad7d1bea23a0614752418/tests/hardware_reboot/testdata/hardware_usb_reboot.txt#L33

File: tests/hardware_reboot/testdata/hardware_usb_reboot.txt [hunk 2]
```diff
@@ -61,19 +83,19 @@ do
...
 
--- get-lshw.sh --
+-- get-last-reboot-time.sh --
 port=$1
 EDEN={{EdenConfig "eden.root"}}/{{EdenConfig "eden.bin-dist"}}/{{EdenConfig "eden.eden-bin"}}
 HOST=$($EDEN eve ip)
- echo ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST lshw
- ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST lshw
+ echo ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST 'last -n 1 --time-format iso reboot'
+ ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST 'last -n 1 --time-format iso reboot'
```
Comment (eriknordmark) [inline tests/hardware_reboot/testdata/hardware_usb_reboot.txt:98]: This returns a bunch of unrelated stuff in addition to the timestamp.
Since those might differ (who knows), it should be best to extract just the timestamp e.g., using
last -n 1 --time-format iso reboot | head -1 | awk '{print $5}'

**Comment by dautovri** [comment]
@sadov  @eriknordmark 
I moved the test from docker/eclient to ubuntu VM and I have discovered two interesting behavior related to Xen. KVM works fine. 

1) Sometimes app stuck in HALTED state
2) Test is not passing because USB disappears. (repeatable)

```
[diff -before_reboot +after_reboot]
         ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i /home/r/GolandProjects/eden/dist/tests/eclient/image/cert/id_rsa -p 8027 ubuntu@192.168.31.226 lsusb
        -Bus 001 Device 002: ID 46f4:0001 QEMU QEMU USB HARDDRIVE
         Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
        FAIL: testdata/hardware_usb_reboot.txt:41: before_reboot and after_reboot differ
```

How to repeat the test: 
gh pr checkout 560
make build-test 
./eden config add
./eden config set default --key eve.hv --value xen
./eden setup
./eden start && ./eden eve onboard 
./qemu+usb 
./eden eve stop 
./eden eve start 
./eden test tests/hardware_reboot -v debug

File: tests/hardware_reboot/testdata/hardware_usb_reboot.txt [hunk 3]
```diff
@@ -57,27 +66,27 @@ for i in `seq 20`
 do
  sleep 20
  # Test SSH-access to container
- echo $i\) ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST grep Ubuntu /etc/issue
- ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port root@$HOST grep Ubuntu /etc/issue && break
+ echo $i\) ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o PasswordAuthentication=no -i {{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa -p $port ubuntu@$HOST grep Ubuntu /etc/issue
```
Comment (giggsoff) [inline tests/hardware_reboot/testdata/hardware_usb_reboot.txt:69]: Please change all lines from `{{EdenConfig "eden.root"}}/tests/eclient/image/cert/id_rsa` to `{{EdenConfig "eden.tests"}}/eclient/image/cert/id_rsa`

**Comment by zed-sadov** [comment]
Seems like some typos and misguided steps. My sequence of commands to execute the test:
```
make build-tests 
./eden config set default --key eve.hv --value xen
./eden setup
./dist/bin/qemu+usb.sh 
./dist/bin/eden+ports.sh 2223:2223
./eden start && ./eden eve onboard 
./eden test tests/hardware_reboot -v debug
```
The result appears to be as described -- test log:
[hardware_reboot.001.log](https://github.com/lf-edge/eden/files/6244484/hardware_reboot.001.log)
.

File: tests/hardware_reboot/testdata/hardware_usb_reboot.txt (hunk 3 above)
Comment (dautovri) [reply tests/hardware_reboot/testdata/hardware_usb_reboot.txt:69]: Ok, I will change.

**Review by dautovri — COMMENTED** [review]
Signed-off-by: Ruslan Dautov dautov2@gmail.com

**Comment by giggsoff** [comment]
In `xenstore-ls` I see before reboot from inside:
```
vusb = ""
    0 = ""
     type = "devicemodel"
     usb-ver = "2"
     num-ports = "6"
     port = ""
      1 = "2-1"
      2 = ""
      3 = ""
      4 = ""
      5 = ""
      6 = ""
```
After:
```
   vusb = ""
    0 = ""
     type = "devicemodel"
     usb-ver = "2"
     num-ports = "6"
     port = ""
      1 = ""
      2 = ""
      3 = ""
      4 = ""
      5 = ""
      6 = ""
```
So, seems, that usb devices not persist reboot from inside die to behavior of Xen.
